File: native-engine/bench.ts
/* Matched TS micro-benchmark — same computation as bench.cpp. pnpm tsx native-engine/bench.ts [N] */
import { Cyclotomic, CyclotomicRing, setActiveRing } from "@/classes/Cyclotomic";

const N = Number(process.argv[2] ?? 2000000);
const R = CyclotomicRing.create(24);
setActiveRing(R);
const C = (n: number[], d: number) => new Cyclotomic(R, n.map(BigInt), BigInt(d));
const x = C([3, -1, 4, 1, -5, 9, -2, 6], 7);
const y = C([1, 2, -3, 0, 5, -1, 2, 4], 3);
const z = C([-2, 5, 1, -4, 0, 3, -1, 2], 5);

let checksum = 0n;
const t0 = performance.now();
for (let i = 0; i < N; i++) {
  const t = x.mulZeta(i % 24).mul(y).add(z).conj();
  checksum += t.num[0] + t.den;
}
const ms = performance.now() - t0;
console.log(`TS:     N=${N}  ${ms.toFixed(1)} ms  ${(N / ms / 1000).toFixed(2)} Mops/s  checksum=${checksum}`);

File: native-engine/bench.cpp
// Matched micro-benchmark: N iterations of an independent mulZeta→mul→add→conj chain in native
// int128 exact arithmetic. Prints wall time + a checksum (must equal the TS bench's checksum → a
// bonus correctness cross-check). Pair with bench.ts (same computation).
#include <cstdio>
#include <chrono>
#include "cyclotomic.hpp"

int main(int argc, char** argv) {
    long N = argc > 1 ? atol(argv[1]) : 2000000;
    const Ring* R = &Ring::create(24);
    Cyclo x(R, {3, -1, 4, 1, -5, 9, -2, 6}, 7);
    Cyclo y(R, {1, 2, -3, 0, 5, -1, 2, 4}, 3);
    Cyclo z(R, {-2, 5, 1, -4, 0, 3, -1, 2}, 5);
    i128 checksum = 0;
    auto t0 = std::chrono::steady_clock::now();
    for (long i = 0; i < N; i++) {
        Cyclo t = x.mulZeta((int)(i % 24)).mul(y).add(z).conj();
        t.ensureCanon();  // deferred gcd ⇒ canonicalize before reading num/den (matches TS/bench-fast checksum)
        checksum += t.num[0] + t.den;
    }
    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    printf("native: N=%ld  %.1f ms  %.2f Mops/s  checksum=%s\n",
           N, ms, N / ms / 1000.0, to_string128(checksum).c_str());
    return 0;
}
